## COMP 7118 Assignment 2
### DUE: 9/27/2018

#### You are supposed to do this assignment by yourself.  You can discuss with your classmates.  But Python code must be written by you.  Read the plagiarism policy in the syllabus for additional information




### Rationale

In this assignment, you are presented with a fictional scenario, in which you need to build a Bloom filter to detect dangerous viral DNAs, given limited amount of storage.

Imagine you have to build a device with DNA sensors. The device is deployed in remote location without much power (other than solar) and resources (not a lot of computing power and RAM).

The device constantly gets DNA inputs from the environment via its sensors.  And if it detects a known dangerous virus, it will send a signal back to the control station, alerting people of the danger.

It is best for this assignment that you write your Python code in an external file, and demonstrate how to use it in this Jupyter notebook.

## YOUR NAME: Md Lutfar Rahman

**(1)** Write a Python class called BloomFilter, with a constructor that takes as input a list of n DNA strings (of the virus) and a desirable false positive error probability p.  Assume that each string has a length L.

Your task is to derive the number of bits, m, and the number of hash functions, k, to the strings and achieve the given false positive probability.

Show your calculation and Python code.




Note on optimal parameters for Bloom filters. Given n items and an m-bit array, the optimal value for the number of hash functions is $k = {m \over n} \cdot ln(2)$.  This results in a false positive probability of approximately ${1 \over 2}^{{m \over n} \cdot ln(2)}$.

You can use the bitarray module that comes with Anaconda.  See below:




## ANS(1): 
number of bits, $m = {n\ln{p} \over {\ln{2} \ln{0.5}}} $  
the number of hash functions, $k = {m \over n} \cdot ln(2)$  

In [1]:
#ANS(1):(code)
from solution import *
n = 1000
L = 100
p = .20
DNAs = generate_random_dnas(n, L)

bloomfilter = BloomFilter(DNAs, p)
m, k = bloomfilter.get_m_k(n, p)

print(m,k)

3349 2


**(2)** Write a function that generates a random DNA of length L. Demonstrate how to use this method.

## ANS(2)

In [2]:
#ANS(2)
L = 10
print(generate_random_dna(L))

GAATATAAGA


**(3)** Write a method that generates k random hash functions for DNA strings of length L.  Before you do that, you must describe in English clearly how to construct each hash function.  The idea is to map a string of characters A, C, G, and T (think of them as 0, 1, 2, and 3) to a random number between 1 and m.

This problem requires some reasoning, not just coding.

## ANS(3)
I converted each dna to a number by considering A, C, G, and T as 0,1,2,3. Then converted it as a decimal number. The idea is same as in our assignment 1. Then this number is passed to the hash functions. The basic sructure of the hash function is same as we did in the class. The only difference here is, I needed to fix different snapshots of a and b for each hash functions.

In [3]:
#ANS(3)
hashes = bloomfilter.generate_k_hash_functions(k, m)
hashes[0]('ACCCATGCCA')

672

**(4)** Write a method that inserts a DNA string of length L into the Bloom filter.  Demonstrate how to use this method with actual Python code.

## ANS(4)

In [4]:
#ANS (4)
dna = 'ACGCTTGCCA'
bloomfilter.prepareBloomFilter()
bloomfilter.insert(dna)
print(dna, 'inserted')

ACGCTTGCCA inserted


**(5)** Write a method that checks if an unknown DNA string of length L is a viral DNA.  Demonstrate how to use this method with actual Python code.

## ANS(5)

In [5]:
#ANS(5)
dna = 'ACGCTTGCCA'
print(bloomfilter.filterCheck(dna))

True


**(6)** Write an evaluation function (not part of the BloomFilter class) that generates a  viral genome containing 10,000 random DNA strings of length 100.

Generate N random DNA strings of length 100, and test to see if your bloom filter work.  In particular, report the false positive probability (which is the number of false positives divided by N).

To check if a query is a false positive, you should store separate those 1 millions DNA strings (viral genome) as ground truth.  The Bloom Filter won't be able to tell you for sure if a "Yes" query is in fact a yes.

Demonstrate how to use this function.



## ANS(6)

In [6]:
#ANS(6)
N = 500
n = 10000
L = 100
p = .20
evaluate(N, n, L, p)

0.348


Observation: .20 and .348 looks close enough! A better hash function might improve it.